### PR TITLE
Extract metrics API that is decoupled from dropwizard [PLEN-28]

### PR DESCRIPTION
**Pull request by nicu-da**
This will allow us to add another implementation for metrics in the future (OpenTelemetry) while keeping the API unchanged.
In a follow up PR some changes to the API will be done to add support for labels in the API itself.

### CODE REVIEW

**Comment by atriantafyllos-da** [comment]
As a first comment this will break canton. I must try to compile with canton to see the conflicts.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:28]: removing `addGauge` will break canton in 
https://github.com/DACH-NY/canton/blob/e888db5dc5ab7dccba61ebef171fb2e5c46b4d3d/community/common/src/main/scala/com/digitalasset/canton/metrics/MetricHandle.scala#L30-L36

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:98]: This change also breaks canton e.g. https://github.com/DACH-NY/canton/blob/e888db5dc5ab7dccba61ebef171fb2e5c46b4d3d/community/common/src/main/scala/com/digitalasset/canton/crypto/store/db/DbCryptoPrivateStore.scala#L70.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:129]: Needed by canton.

File: ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala:63]: this function must be exported for canton

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (atriantafyllos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:87]: breaks canton if removed

**Review by atriantafyllos-da — CHANGES_REQUESTED** [review]

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:28]: Based on what I'm seeing in canton, the `RefGauge` is more or less a `VarGauge`. The metrics using `RefGauge` should in fact use a `gaugeWithSupplier`, and the `RefGuage` should be deleted.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:98]: We're mixing concepts in a pretty difficult way here. The `TimedLoadGoage` should not exist, instead this should be an extension for `Timer` that adds cats ways to work with cats.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:129]: We should not depend on the internals of the metrics themselves. If required for testing we can create assertions that work with metrics.

File: ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/Timed.scala:63]: It depends on `Dropwizard`, canton should not use dropwizard metrics directly.

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala
Comment (nicu-da) [reply ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala:87]: It should be moved together with themethods from the `TimedLoadGauge` as it seems to target the same API.

**Comment by nicu-da** [comment]
> As a first comment this will break canton. I must try to compile with canton to see the conflicts.

@atriantafyllos-da Thank you very much for checking the integration in canton. I think you discovered some good spots where it would break, and based on what I've seen from the usages we removed I think the changes should happen in canton itself, so adapt to the simplified API.

**Review by simonmaxen-da — COMMENTED** [review]
LGTM

**Review by atriantafyllos-da — APPROVED** [review]
Approved but please do not merge before the PR for canton is approved.

**Pull request merged**
